File: src/src/gestorAplicacion/servicios/Usuario.java
package gestorAplicacion.servicios;

import gestorAplicacion.libreria.CATEGORIA;
import gestorAplicacion.libreria.GENERO;
import gestorAplicacion.libreria.Libro;
import gestorAplicacion.libreria.Revista;

import java.io.Serializable;
import java.util.*;

public class Usuario implements Serializable {
    private String nombre;
    private int id;
    private ArrayList<Prestamo> prestamos = new ArrayList<>();
    private ArrayList<Reserva> reservas = new ArrayList<>();
    private ArrayList<Tiquete> tiquetes = new ArrayList<>();
    private ArrayList<Libro> historialLibrosUsados = new ArrayList<>();
    private ArrayList<Revista> historialRevistasUsadas = new ArrayList<>();
    private boolean multa = false;
    private GENERO generoFavorito;
    private CATEGORIA categoriaFavorita;

    //constructor

    public Usuario(String nombre, int id) {
        this.nombre = nombre;
        this.id = id;
    }

    //getters y setters
    public String getNombre() {
        return nombre;
    }

    public void setNombre(String nombre) {
        this.nombre = nombre;
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public ArrayList<Prestamo> getPrestamos() {
        return prestamos;
    }

    public void setPrestamos(ArrayList<Prestamo> prestamos) {
        this.prestamos = prestamos;
    }

    public ArrayList<Reserva> getReservas() {
        return reservas;
    }

    public void setReservas(ArrayList<Reserva> reservas) {
        this.reservas = reservas;
    }

    public ArrayList<Tiquete> getTiquetes() {
        return tiquetes;
    }

    public void setTiquetes(ArrayList<Tiquete> tiquetes) {
        this.tiquetes = tiquetes;
    }

    public boolean isMulta() {
        return multa;
    }

    public void setMulta(boolean multa) {
        this.multa = multa;
    }

    public GENERO getGeneroFavorito() {
        return generoFavorito;
    }

    public void setGeneroFavorito(GENERO generoFavorito) {
        this.generoFavorito = generoFavorito;
    }

    public ArrayList<Libro> getHistorialLibrosUsados() {
        return historialLibrosUsados;
    }

    public void setHistorialLibrosUsados(ArrayList<Libro> historialLibrosUsados) {
        this.historialLibrosUsados = historialLibrosUsados;
    }

    public ArrayList<Revista> getHistorialRevistasUsadas() {
        return historialRevistasUsadas;
    }

    public void setHistorialRevistasUsadas(ArrayList<Revista> historialRevistasUsadas) {
        this.historialRevistasUsadas = historialRevistasUsadas;
    }

    public CATEGORIA getCategoriaFavorita() {
        return categoriaFavorita;
    }

    public void setCategoriaFavorita(CATEGORIA categoriaFavorita) {
        this.categoriaFavorita = categoriaFavorita;
    }

    //metodos

    /**
     *
     */
    public void encontrarGeneroFavorito(){
        // hashmap para guardar la frecuencia de lectura de genero
        Map<GENERO, Integer> generosLeidos = new HashMap<GENERO, Integer>();
        GENERO generoFavorito = null;
        int maxLeido = 0;

        for (Libro libro : historialLibrosUsados) {
            Integer j = generosLeidos.get(libro.getGenero());
            generosLeidos.put(libro.getGenero(), (j == null) ? 1 : j + 1);
        }


        //se busca el genero favorito con el hashMap
        for (GENERO genero: generosLeidos.keySet()) {
            if(generosLeidos.get(genero)>maxLeido){
                generoFavorito = genero;
                maxLeido = generosLeidos.get(genero);
            }
        }
        setGeneroFavorito(generoFavorito);

    }

    public void encontrarCategoriaFavorita(){
        // hashmap para guardar la frecuencia de lectura de genero
        Map<CATEGORIA, Integer> categoriasLeidas = new HashMap<CATEGORIA, Integer>();
        CATEGORIA categoriaFavorita = null;
        int maxLeido = 0;

        for (Revista revista : historialRevistasUsadas) {
            Integer j = categoriasLeidas.get(revista.getCategoria());
            categoriasLeidas.put(revista.getCategoria(), (j == null) ? 1 : j + 1);
        }



        //se busca el genero favorito con el hashMap
        for (CATEGORIA categoria: categoriasLeidas.keySet()) {
            if(categoriasLeidas.get(categoria)>maxLeido){
                categoriaFavorita = categoria;
                maxLeido = categoriasLeidas.get(categoria);
            }
        }
        setCategoriaFavorita(categoriaFavorita);

    }
}

File: Python/gestorAplicacion/servicios/usuario.py
from Python.gestorAplicacion.libreria.categoria import CATEGORIA
from Python.gestorAplicacion.libreria.genero import GENERO
from Python.gestorAplicacion.libreria.libro import Libro
from Python.gestorAplicacion.libreria.revista import Revista
from Python.gestorAplicacion.servicios.prestamo import Prestamo
from Python.gestorAplicacion.servicios.reserva import Reserva
from Python.gestorAplicacion.servicios.tiquete import Tiquete


class Usuario():
    def __init__(self, nombre: str, id: int, prestamos: list, reservas: list, tiquetes: list, historialLibrosUsados: list,
                 historialResvistasUsadas: list, multa = False, generoFavorito = GENERO.NOVELA, categoriaFavorita = CATEGORIA.ACTUALIDAD):
        self._nombre = nombre
        self._id = id
        self._prestamos = prestamos
        self._reservas = reservas
        self._tiquetes = tiquetes
        self._historialLibrosUsados = historialLibrosUsados
        self._historialRevistasUsadas = historialResvistasUsadas
        self._multa = multa
        self._generoFavorito = generoFavorito
        self._categoriaFavorita = categoriaFavorita

    #getters y setters
    def getNombre(self)-> str:
        return self._nombre

    def getId(self) -> int:
        return self._id

    def getPrestamos(self)-> list:
        return self._prestamos

    def getReservas(self)-> list:
        return self._reservas

    def getTiquetes(self)-> list:
        return self._tiquetes

    def getHistorialLibrosUsados(self)-> list:
        return self._historialLibrosUsados

    def getHistorialRevistasUsadas(self)-> list:
        return self._historialRevistasUsadas

    def getGeneroFavorito(self)-> GENERO:
        return self._generoFavorito

    def getCategoriaFavorita(self)-> CATEGORIA:
        return self._categoriaFavorita

    def setNombre(self, nombre: str):
        self._nombre = nombre

    def setId(self, id: int):
        self._id = id

    def setPrestamos(self, prestamos: list):
        self._prestamos = prestamos

    def setReservas(self, reservas: list):
        self._reservas = reservas

    def setTiquetes(self, tiquetes: list):
        self._tiquetes = tiquetes

    def setHistorialLibrosUsados(self, historial: list):
        self._historialLibrosUsados = historial

    def setHistorialRevistasUsadas(self, historial: list):
        self._historialRevistasUsadas = historial

    def setGeneroFavorito(self, genero = GENERO):
        self._generoFavorito = genero

    def setCategoriaFavorita(self, categoria: CATEGORIA):
        self._categoriaFavorita = categoria

    #Metodos
    """* Se recorre la lista 'historialLibrosUsados' para posteriormente crear un diccionario que contendra 
       la cantidad de veces que se repite cada genero. Teniendo ya este diccionario creado, se busca cu??l
        es el genero que m??s aparece y este se define como el favorito
        """
    def encontrarGeneroFavorito(self):
        generosLeidos = {}
        for libro in self.getHistorialLibrosUsados():
            if libro.getGenero() in generosLeidos:
                generosLeidos[libro.getGenero()] += 1
            else:
                generosLeidos[libro.getGenero()] = 1

        #Buscar el genero m??s leido
        generoFavorito = max(generosLeidos, key=generosLeidos.get, default=0)
        self.setGeneroFavorito(generoFavorito)

    """* Se recorre la lista 'historialRevistasUsadas' para posteriormente crear un diccionario que contendra
       la cantidad de veces que se repite cada categoria. Teniendo ya este diccionario creado, se busca 
       cu??l es la categoria que m??s aparece y esta se define como la favorita
       """
    def encontrarCategoriaFavorita(self):
        categoriasLeidas = {}
        for revista in self.getHistorialRevistasUsadas():
            if revista.getCategoria() in categoriasLeidas:
                categoriasLeidas[revista.getCategoria()] += 1
            else:
                categoriasLeidas[revista.getCategoria()] = 1

        #Buscar categoria favorita
        categoriaFavorita = max(categoriasLeidas, key=categoriasLeidas.get, default=0)
        self.setCategoriaFavorita(categoriaFavorita)

    def a??adirHistorialLibros(self, libro: Libro):
        self._historialLibrosUsados.append(libro)

    def a??adirHistorialRevistas(self, revista: Revista):
        self._historialRevistasUsadas.append(revista)

    def a??adirReserva(self, reserva: Reserva):
        self._reservas.append(reserva)

    def a??adirPrestamo(self, prestamo: Prestamo):
        self._prestamos.append(prestamo)

    def a??adirTiquete(self, tiquete: Tiquete):
        self._tiquetes.append(tiquete)

    def eliminarReserva(self, reserva: Reserva):
        for r in self._reservas:
            if reserva == r:
                self._reservas.remove(reserva)

    def eliminarPrestamo(self, prestamo: Prestamo):
        for p in self._prestamos:
            if prestamo == p:
                self._prestamos.remove(prestamo)

    def eliminarTiquete(self, tiquete: Tiquete):
        self._tiquetes.remove(tiquete)



